checker/src/org/checkerframework/checker/index/upperbound/UpperBoundVisitor.java

```diff
@@ -33,6 +52,150 @@ public UpperBoundVisitor(BaseTypeChecker checker) {
+     * or effectively final are always permitted in dependent annotations, even in local variables,
+     * and will not trigger this warning (as no side-effect can effect them).
+     *
+     * <p>Also issues warnings if the expression in a dependent annotation is not permitted.
+     * Permitted expressions in dependent annotations must be at least one of the following:
+     *
+     * <ul>
+     *   <li>final or effectively final variables
+     *   <li>local variables
+     *   <li>private fields
+     *   <li>pure method calls all of whose arguments (including the receiver expression) are
+     *       composed only of expressions in this list
+     *   <li>accesses of a public final field whose access expression (sometimes called the
+     *       receiver) is composed only of expressions in this list
+     * </ul>
+     *
+     * <p>Any other expression results in a warning.
+     */
+    @Override
+    public Void visitVariable(VariableTree node, Void p) {
+        AnnotatedTypeMirror atm = atypeFactory.getAnnotatedTypeLhs(node);
+        AnnotationMirror anm = atm.getAnnotationInHierarchy(atypeFactory.UNKNOWN);
+        if (anm != null && AnnotationUtils.hasElementValue(anm, "value")) {
+            // This is a dependent annotation. If this is a local variable,
+            // issue a warning; dependent annotations should not be written on local variables.
+            Element elt = TreeUtils.elementFromDeclaration(node);
+            boolean allVariablesFinal = true;
+
+            UBQualifier qual = UBQualifier.createUBQualifier(anm);
+            if (qual.isLessThanLengthQualifier()) {
+                LessThanLengthOf ltl = (LessThanLengthOf) qual;
+                for (String array : ltl.getSequences()) {
+                    Receiver rec = getReceiverForCheckingDependentTypes(array, node);
+                    // covers final and effectively final variables
+                    boolean isFinal =
+                            rec == null
+                                    || rec.isUnmodifiableByOtherCode()
+                                    || rec instanceof FlowExpressions.LocalVariable;
+                    if (!isFinal) {
+                        allVariablesFinal = false;
+                        checkIfPermittedInDependentTypeAnno(rec, array, node);
+                    }
+                }
+            }
+
+            if (elt.getKind() == ElementKind.LOCAL_VARIABLE && !allVariablesFinal) {
+                checker.report(Result.warning(LOCAL_VAR_ANNO), node);
+            }
+        }
+
+        return super.visitVariable(node, p);
+    }
+
+    private Receiver getReceiverForCheckingDependentTypes(String expr, Tree tree) {
+        try {
+            return atypeFactory.getReceiverFromJavaExpressionString(expr, getCurrentPath());
+        } catch (FlowExpressionParseUtil.FlowExpressionParseException e) {
+            // issue warning
+            checker.report(
```

Comment:
If the expression isn't parsable, then shouldn't the checker already have issued a warning?  Is this really a duplicate warning?

Reply:
I don't think so. `FlowExpressionParseUtil#parse` and its helper methods just throw an exception, and don't issue a checker warning. So catching this is appropriate.

Reply:
The exception needs to be caught, but BaseTypeVisitor also parses the expression and issues a warning `expression.unparsable.type.invalid` warning.  For example:
```
import org.checkerframework.checker.index.qual.IndexFor;

class Unparsable {
   //::error: (expression.unparsable.type.invalid)
   @IndexFor("sdlfj") int x;
}
```

Reply:
Makes sense. I'll remove the extra warning.